db/sqldb/analysis.go

```diff
@@ -0,0 +1,237 @@
+
+func (d *DB) Analysis(tx db.Tx, id pacta.AnalysisID) (*pacta.Analysis, error) {
+	rows, err := d.query(tx, `
+		SELECT `+analyseSelectColumns+`
+		FROM analysis 
+		WHERE id = $1;`, id)
+	if err != nil {
+		return nil, fmt.Errorf("querying analysis: %w", err)
+	}
+	pvs, err := rowsToAnalyses(rows)
+	if err != nil {
+		return nil, fmt.Errorf("translating rows to analyses: %w", err)
+	}
+	return exactlyOne("analysis", id, pvs)
+}
+
+func (d *DB) Analyses(tx db.Tx, ids []pacta.AnalysisID) (map[pacta.AnalysisID]*pacta.Analysis, error) {
+	ids = dedupeIDs(ids)
+	rows, err := d.query(tx, `
+		SELECT `+analyseSelectColumns+`
+		FROM analysis 
+		WHERE id IN `+createWhereInFmt(len(ids))+`;`, idsToInterface(ids)...)
+	if err != nil {
+		return nil, fmt.Errorf("querying analyses: %w", err)
+	}
+	pvs, err := rowsToAnalyses(rows)
+	if err != nil {
+		return nil, fmt.Errorf("translating rows to analyses: %w", err)
+	}
+	result := make(map[pacta.AnalysisID]*pacta.Analysis, len(pvs))
+	for _, pv := range pvs {
+		result[pv.ID] = pv
+	}
+	return result, nil
+}
+
+func (d *DB) CreateAnalysis(tx db.Tx, a *pacta.Analysis) (pacta.AnalysisID, error) {
+	if err := validateAnalysisForCreation(a); err != nil {
+		return "", fmt.Errorf("validating analysis for creation: %w", err)
+	}
+	a.ID = pacta.AnalysisID(d.randomID("analysis"))
+	err := d.exec(tx, `
+		INSERT INTO analysis 
+			(id, analysis_type, owner_id, pacta_version_id, portfolio_snapshot_id, name, description)
+			VALUES
+			($1, $2, $3, $4, $5, $6, $7);`,
+		a.ID, a.AnalysisType, a.Owner.ID, a.PACTAVersion.ID, a.PortfolioSnapshot.ID, a.Name, a.Description)
+	if err != nil {
+		return "", fmt.Errorf("creating analysis: %w", err)
+	}
+	return a.ID, nil
+}
+
+func (d *DB) UpdateAnalysis(tx db.Tx, id pacta.AnalysisID, mutations ...db.UpdateAnalysisFn) error {
+	err := d.RunOrContinueTransaction(tx, func(db.Tx) error {
+		init, err := d.Analysis(tx, id)
+		if err != nil {
+			return fmt.Errorf("reading analysis: %w", err)
+		}
+		for i, m := range mutations {
+			err := m(init)
+			if err != nil {
+				return fmt.Errorf("running %d-th mutation: %w", i, err)
+			}
+		}
+		err = d.putAnalysis(tx, init)
+		if err != nil {
+			return fmt.Errorf("putting analysis: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("updating analysis: %w", err)
+	}
+	return nil
+}
+
+func (d *DB) DeleteAnalysis(tx db.Tx, id pacta.AnalysisID) ([]pacta.BlobURI, error) {
+	buris := []pacta.BlobURI{}
+	err := d.RunOrContinueTransaction(tx, func(tx db.Tx) error {
+		a, err := d.Analysis(tx, id)
+		if err != nil {
+			return fmt.Errorf("reading analysis: %w", err)
+		}
+		aas, err := d.AnalysisArtifactsForAnalysis(tx, id)
+		if err != nil {
+			return fmt.Errorf("reading analysis artifacts: %w", err)
+		}
+		for _, aa := range aas {
+			buri, err := d.DeleteAnalysisArtifact(tx, aa.ID)
+			if err != nil {
+				return fmt.Errorf("deleting analysis artifact %q: %w", aa.ID, err)
+			}
+			buris = append(buris, buri)
+		}
```

Comment:
Ah - now I remember why. It's because `blob_uri` is a feature on `blob` not on `artifact`. This does the trick though.

```
                 rows, err := d.query(tx, `
			WITH deleted_analysis_artifacts AS (
				DELETE FROM analysis_artifact
				WHERE analysis_id = your_analysis_id_here
				RETURNING blob_id
			)
			DELETE FROM blob
			WHERE id IN (SELECT blob_id FROM deleted_analysis_artifacts)
			RETURNING blob_uri;
		`, id)
		if err != nil {
			return fmt.Errorf("reading analysis_artifacts: %w", err)
		}
```